### PR TITLE
Adds partition lost listener configuration to ICache.

### PR DESCRIPTION
**Pull request by kobalski**

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/cache/ICache.java [hunk 1]
```diff
@@ -944,4 +945,31 @@
...
+     * The addPartitionLostListener returns a register-id. This id is needed to remove the MapPartitionLostListener using the
+     * {@link #removePartitionLostListener(String)} method.
+     * <p/>
+     * There is no check for duplicate registrations, so if you register the listener twice, it will get events twice.
+     * IMPORTANT: Please @see com.hazelcast.partition.PartitionLostListener for weaknesses.
+     * IMPORTANT: Listeners registered from HazelcastClient may miss some of the cache partition lost events due
+     * to design limitations.
+     *
+     * @param listener the added MapPartitionLostListener.
+     * @return returns the registration id for the MapPartitionLostListener.
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/cache/ICache.java]: map or cache? :)

Comment (kobalski) [inline hazelcast/src/main/java/com/hazelcast/cache/ICache.java]: This is why copy paste programming is so wrong :)

File: hazelcast/src/main/java/com/hazelcast/cache/ICache.java [hunk 2]
```diff
@@ -944,4 +945,31 @@
...
+    String addPartitionLostListener(CachePartitionLostListener listener);
+
+    /**
+     * Removes the specified cache partition lost listener.
+     * Returns silently if there is no such listener added before.
+     *
+     * @param id id of registered listener.
+     * @return true if registration is removed, false otherwise.
+     */
+    boolean removePartitionLostListener(String id);
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/cache/ICache.java]: you are throwing NPE when id is `null`.. Can you also write this info to javadoc?

File: hazelcast/src/main/java/com/hazelcast/config/CachePartitionLostListenerConfigReadOnly.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    public ListenerConfig setClassName(String className) {
+        throw new UnsupportedOperationException("this config is read-only");
+    }
+
+    public ListenerConfig setImplementation(EventListener implementation) {
+        throw new UnsupportedOperationException("this config is read-only");
+    }
+
+    public MapPartitionLostListenerConfig setImplementation(MapPartitionLostListener implementation) {
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/config/CachePartitionLostListenerConfigReadOnly.java]: i guess this should be `CachePartitionLostListenerConfig`

File: hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheAddPartitionLostListenerMessageTask.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    @Override
+    public String getMethodName() {
+        return "addCachePartitionLostListener";
+    }
+
+    @Override
+    public Object[] getParameters() {
+        return null;
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheAddPartitionLostListenerMessageTask.java]: why this is `null`?

Comment (kobalski) [inline hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/cache/CacheAddPartitionLostListenerMessageTask.java]: To indicate that it is not used.

**Comment by metanet** [comment]
:+1:

**Pull request merged**